File: services/anomaly-detector/metrics.go
package main

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	metricMessagesConsumed = promauto.NewCounter(prometheus.CounterOpts{
		Name: "messages_consumed_total",
		Help: "Telemetry messages consumed from telemetry.processed.",
	})

	metricAnomaliesDetected = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "anomalies_detected_total",
		Help: "Anomalies detected, by method.",
	}, []string{"method"})

	metricMessagesFailed = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "messages_failed_total",
		Help: "Messages that failed processing, by reason.",
	}, []string{"reason"})

	metricDLQMessages = promauto.NewCounter(prometheus.CounterOpts{
		Name: "dlq_messages_total",
		Help: "Messages routed to the dead-letter topic.",
	})

	metricProcessingLatency = promauto.NewHistogram(prometheus.HistogramOpts{
		Name:    "processing_latency_seconds",
		Help:    "Time from Kafka fetch to commit.",
		Buckets: prometheus.DefBuckets,
	})

	metricKafkaConsumerLag = promauto.NewGauge(prometheus.GaugeOpts{
		Name: "kafka_consumer_lag",
		Help: "Consumer lag on telemetry.processed.",
	})

	metricForestsTrained = promauto.NewCounter(prometheus.CounterOpts{
		Name: "isolation_forests_trained_total",
		Help: "Isolation forest (re)training runs completed, across all machine types.",
	})
)

File: services/anomaly-detector/metrics.py
"""Prometheus metrics, the Python port of metrics.go. Same metric names and
labels, so existing Grafana dashboards/alerts keep working unchanged."""

from prometheus_client import Counter, Gauge, Histogram

messages_consumed_total = Counter(
    "messages_consumed_total",
    "Telemetry messages consumed from telemetry.processed.",
)

anomalies_detected_total = Counter(
    "anomalies_detected_total",
    "Anomalies detected, by method.",
    ["method"],
)

messages_failed_total = Counter(
    "messages_failed_total",
    "Messages that failed processing, by reason.",
    ["reason"],
)

dlq_messages_total = Counter(
    "dlq_messages_total",
    "Messages routed to the dead-letter topic.",
)

processing_latency_seconds = Histogram(
    "processing_latency_seconds",
    "Time from Kafka fetch to commit.",
)

kafka_consumer_lag = Gauge(
    "kafka_consumer_lag",
    "Consumer lag on telemetry.processed.",
)

isolation_forests_trained_total = Counter(
    "isolation_forests_trained_total",
    "Isolation forest (re)training runs completed, across all machine types.",
)
